go/chat/flipmanager.go

```diff
@@ -0,0 +1,736 @@
+	if err != nil {
+		return res, err
+	}
+	if !msg.IsValid() {
+		return res, errors.New("host message invalid")
+	}
+	if !msg.Valid().MessageBody.IsType(chat1.MessageType_TEXT) {
+		return res, fmt.Errorf("invalid host message type: %v", msg.GetMessageType())
+	}
+	body := msg.Valid().MessageBody.Text().Body
+	if err := json.Unmarshal([]byte(body), &res); err != nil {
+		return res, err
+	}
+	return res, nil
+}
+
+// StartFlip implements the types.CoinFlipManager interface
+func (m *FlipManager) StartFlip(ctx context.Context, uid gregor1.UID, hostConvID chat1.ConversationID,
+	tlfName, text string) (err error) {
+	defer m.Trace(ctx, func() error { return err }, "StartFlip: convID: %s", hostConvID)()
+	gameID := flip.GenerateGameID()
+	m.Debug(ctx, "StartFlip: using gameID: %s", gameID)
+
+	// Get host conv using local storage, just bail out if we don't have it
+	hostConv, err := utils.GetUnverifiedConv(ctx, m.G(), uid, hostConvID,
+		types.InboxSourceDataSourceLocalOnly)
+	if err != nil {
+		return err
+	}
+
+	// First generate the message representing the flip into the host conversation. We also wait for it
+	// to actually get sent before doing anything flip related.
+	outboxID, err := storage.NewOutboxID()
+	if err != nil {
+		return err
+	}
+	listener := newSentMessageListener(m.G(), outboxID)
+	nid := m.G().NotifyRouter.AddListener(listener)
+	if _, err = m.G().ChatHelper.SendMsgByIDNonblock(ctx, hostConvID, tlfName,
+		chat1.NewMessageBodyWithText(chat1.MessageText{
+			Body:       text,
+			FlipGameID: &gameID,
+		}), chat1.MessageType_TEXT, &outboxID); err != nil {
+		return err
+	}
+	sendRes := <-listener.listenCh
+	if sendRes.Err != nil {
+		return sendRes.Err
+	}
+	m.G().NotifyRouter.RemoveListener(nid)
+
+	// Generate dev channel for game messages
+	topicName := m.gameTopicNameFromGameID(gameID)
+	conv, err := m.G().ChatHelper.NewConversation(ctx, uid, tlfName, &topicName, chat1.TopicType_DEV,
+		hostConv.GetMembersType(), keybase1.TLFVisibility_PRIVATE)
+	if err != nil {
+		return err
+	}
+
+	// Record metadata of the host message into the game thread as the first message
+	start, lowerBound, shuffleItems := m.startFromText(text)
+	infoBody, err := json.Marshal(hostMessageInfo{
+		ConvID:       hostConvID,
+		MsgID:        sendRes.MsgID,
+		LowerBound:   lowerBound,
+		ShuffleItems: shuffleItems,
+	})
+	if err != nil {
+		return err
+	}
+	if err := m.G().ChatHelper.SendTextByID(ctx, conv.GetConvID(), tlfName, string(infoBody)); err != nil {
+		return err
+	}
+
+	// Start the game
+	return m.dealer.StartFlipWithGameID(ctx, start, conv.GetConvID(), gameID)
+}
+
+// MaybeInjectFlipMessage implements the types.CoinFlipManager interface
+func (m *FlipManager) MaybeInjectFlipMessage(ctx context.Context, msg chat1.MessageUnboxed,
+	conv chat1.ConversationLocal) {
+	// earliest of outs if this isn't a dev convo, an error, or the outbox ID message
+	if conv.GetTopicType() != chat1.TopicType_DEV || !msg.IsValid() || msg.GetMessageID() == 2 {
```

Comment:
I wonder if we should comment this `== 2` better or make it slightly less fragile should anything change in the future